metadata-io/src/main/java/com/linkedin/metadata/graph/neo4j/Neo4jGraphService.java

```diff
@@ -93,6 +105,63 @@ public void addEdge(@Nonnull final Edge edge) {
     executeStatements(statements);
   }
 
+  @Nonnull
+  @Override
+  public EntityLineageResult getLineage(@Nonnull Urn entityUrn, @Nonnull LineageDirection direction,
+                                        GraphFilters graphFilters, int offset, int count, int maxHops) {
+    log.debug(String.format("Neo4j getLineage maxHops = %d", maxHops));
+
+    final String multiHopTemplate = "MATCH (a {urn: '%s'})-[r:%s*1..%d]->(b) WHERE b:%s RETURN a,r,b"
+            + " UNION "
+            + "MATCH (a {urn: '%s'})<-[r:%s*1..%d]-(b) WHERE b:%s RETURN a,r,b";
+
+    final String directRelationTypes = "DownstreamOf|HasOwner|KnowsUser|Consumes";
+    final String invertedRelationTypes = "Produces";
+    String upstreamRel = direction == LineageDirection.UPSTREAM ? directRelationTypes : invertedRelationTypes;
+    String dowStreamRel = direction == LineageDirection.UPSTREAM ? invertedRelationTypes : directRelationTypes;
+    if (direction == LineageDirection.$UNKNOWN) {
+      upstreamRel = directRelationTypes + "|" + invertedRelationTypes;
+      dowStreamRel = invertedRelationTypes + "|" + directRelationTypes;
+    }
+    final String allowedEntityTypes = String.join(" OR b:", graphFilters.getAllowedEntityTypes());
+    final String statement = String.format(multiHopTemplate,
+            entityUrn, upstreamRel, maxHops, allowedEntityTypes,
+            entityUrn, dowStreamRel, maxHops, allowedEntityTypes);
+
+    List<Record> neo4jResult = runQuery(buildStatement(statement, new HashMap<>())).list();
+    LineageRelationshipArray relations = new LineageRelationshipArray();
+    neo4jResult = neo4jResult.stream()
+            .collect(Collectors.toMap(item -> item.values().get(2).asNode().get("urn").asString(),
+                    Function.identity(),
+                    (item1, item2) -> item1.get(1).size() < item2.get(1).size() ? item1 : item2))
+            .values()
+            .stream()
+            .collect(Collectors.toList());
+      neo4jResult.stream()
+              .skip(offset).limit(offset + count)
+              .forEach(item -> {
+      String urn = item.values().get(2).asNode().get("urn").asString();
+      String relationType = ((InternalRelationship) item.get(1).asList().get(0)).type();
+      int numHops = item.get(1).size();
+      try {
+        relations.add(new LineageRelationship()
+                        .setEntity(Urn.createFromString(urn))
+                        .setType(relationType)
+                        .setDegree(numHops));
+      } catch (URISyntaxException ignored) {
+        log.warn(String.format("Can't convert urn = %s, Error = %s", urn, ignored.getMessage()));
+      }
```

Comment:
its seems like theres some overlap here in executing queries and in the other method- I would consider refactoring before merging.

Reply:
I couldn't find any overlapping but I anyway split the function as it was too long.

Reply:
Love splitting methods :-)